devdocs/PIPELINE_REFACTORING_PLAN.md

```diff
@@ -0,0 +1,277 @@
+└─────────────────────────────────────────────────────────────────┘
+                                │
+                                ▼
+┌─────────────────────────────────────────────────────────────────┐
+│                    BUILDING PHASE                              │
+├─────────────────────────────────────────────────────────────────┤
+│ 1. UMM-G Builder                                               │
+│    └─ Apply templates → UMM-G JSON                            │
+│                                                                 │
+│ 2. CNM Builder                                                 │
+│    └─ Apply templates → CNM JSON                              │
+│                                                                 │
+│ 3. Output Organizer                                           │
+│    └─ Prepare file operations → File operation list            │
+└─────────────────────────────────────────────────────────────────┘
+                                │
+                                ▼
+┌─────────────────────────────────────────────────────────────────┐
+│                    WRITING PHASE                               │
+├─────────────────────────────────────────────────────────────────┤
+│ 1. Local File Writer                                           │
+│    └─ Write UMM-G/CNM files → Local filesystem                │
+│                                                                 │
+│ 2. AWS Stager                                                  │
+│    └─ Upload files → S3 staging area                          │
+│                                                                 │
+│ 3. Notification Publisher                                      │
+│    └─ Send CNM messages → Kinesis stream                      │
+└─────────────────────────────────────────────────────────────────┘
+```
+
+### Core Data Structures (Immutable)
+
+```python
+@dataclass(frozen=True)
+class ProcessingDecisions:
+    reader_strategy: ReaderStrategy
+    geometry_strategy: GeometryStrategy
+    template_strategy: TemplateStrategy
+    output_strategy: OutputStrategy
+
+@dataclass(frozen=True)
+class SourceData:
+    collection_metadata: CollectionMetadata
+    premet_data: Optional[PremetData]
+    spatial_data: Optional[SpatialData]
+    science_metadata: ScienceMetadata
+
+@dataclass(frozen=True)
+class ProcessedMetadata:
+    temporal_extent: TemporalExtent
+    spatial_extent: SpatialExtent
+    additional_attributes: dict
+    file_metadata: dict
+
+@dataclass(frozen=True)
+class BuildOutputs:
+    ummg_content: str
+    cnm_content: str
+    file_operations: List[FileOperation]
+```
+
+## User Stories with Behavioral Acceptance Tests
+
+### Story 1: Data Reading Isolation
+**As a** MetGenC developer, **I can** process science data files through dedicated readers that only handle data extraction, **so that** I/O operations are isolated from business logic and data sources can be easily mocked for testing.
```

Comment:
What kind of data are you planning to extract here? We already have dedicated readers (netcdf, csv) but those are too monolithic.